File: vscode/src/views/CreatePrBodyMarkdown.test.ts
import { describe, expect, it } from "vitest";
import { renderPrBodyMarkdown } from "./CreatePrBodyMarkdown";

describe("renderPrBodyMarkdown", () => {
	it("returns empty string for empty or whitespace-only input", () => {
		expect(renderPrBodyMarkdown("")).toBe("");
		expect(renderPrBodyMarkdown("   \n  \n")).toBe("");
	});

	it("renders ATX headings with a +1 level offset and the md-heading class", () => {
		expect(renderPrBodyMarkdown("## Quick recap")).toBe('<h3 class="md-heading">Quick recap</h3>');
		expect(renderPrBodyMarkdown("# Top")).toBe('<h2 class="md-heading">Top</h2>');
	});

	it("caps heading level at 6", () => {
		expect(renderPrBodyMarkdown("###### Deep")).toContain("<h6");
	});

	it("passes whole-line structural HTML through verbatim so folding renders natively", () => {
		const html = renderPrBodyMarkdown(
			["<details>", "<summary><strong>01 · Title</strong></summary>", "<br>", "<blockquote>", "</blockquote>", "</details>"].join(
				"\n",
			),
		);
		expect(html).toContain("<details>");
		expect(html).toContain("<summary><strong>01 · Title</strong></summary>");
		expect(html).toContain("<br>");
		expect(html).toContain("<blockquote>");
		expect(html).toContain("</blockquote>");
		expect(html).toContain("</details>");
	});

	it("renders a fenced code block, escaping its contents", () => {
		const html = renderPrBodyMarkdown("```ts\nconst x = 1 < 2;\n```");
		expect(html).toBe('<pre class="md-code-block"><code>const x = 1 &lt; 2;</code></pre>');
	});

	it("flushes an unterminated fenced code block", () => {
		const html = renderPrBodyMarkdown("```\nline1\nline2");
		expect(html).toBe('<pre class="md-code-block"><code>line1\nline2</code></pre>');
	});

	it("renders a horizontal rule", () => {
		expect(renderPrBodyMarkdown("---")).toBe('<hr class="md-hr" />');
	});

	it("merges consecutive markdown blockquote lines into one block", () => {
		const html = renderPrBodyMarkdown("> Note: line one\n> line two");
		expect(html).toBe('<blockquote class="md-quote">Note: line one<br />line two</blockquote>');
	});

	it("renders an unordered list and closes it when a non-item line follows", () => {
		const html = renderPrBodyMarkdown("- one\n- two\ntail");
		expect(html).toBe(
			'<ul class="md-list"><li>one</li><li>two</li></ul><div class="md-line">tail</div>',
		);
	});

	it("emits a gap for blank lines and a md-line for paragraphs", () => {
		expect(renderPrBodyMarkdown("hello\n\nworld")).toBe(
			'<div class="md-line">hello</div><div class="md-blank"></div><div class="md-line">world</div>',
		);
	});

	it("applies inline bold, italic, code and http links", () => {
		expect(renderPrBodyMarkdown("**bold**")).toContain("<strong>bold</strong>");
		expect(renderPrBodyMarkdown("*em*")).toContain("<em>em</em>");
		expect(renderPrBodyMarkdown("`code`")).toContain('<code class="md-inline-code">code</code>');
		expect(renderPrBodyMarkdown("[docs](https://x.jolli.ai/p)")).toContain(
			'<a class="md-link" href="https://x.jolli.ai/p">docs</a>',
		);
	});

	it("does not treat underscores in identifiers or paths as italic", () => {
		const html = renderPrBodyMarkdown("Touch `file_name_v2.ts` and a_b_c");
		expect(html).not.toContain("<em>");
	});

	it("escapes stray HTML in prose so it cannot inject markup", () => {
		const html = renderPrBodyMarkdown("Danger <img src=x onerror=1> here");
		expect(html).not.toContain("<img");
		expect(html).toContain("&lt;img");
	});

	it("normalizes CRLF line endings", () => {
		expect(renderPrBodyMarkdown("a\r\nb")).toBe('<div class="md-line">a</div><div class="md-line">b</div>');
	});

	it("does not re-apply bold/italic markup found inside an inline code span", () => {
		const html = renderPrBodyMarkdown("Use `**not bold**` here");
		expect(html).toContain('<code class="md-inline-code">**not bold**</code>');
		expect(html).not.toContain("<strong>");
	});

	it("does not re-apply bold/italic markup found inside a link's text or URL", () => {
		const html = renderPrBodyMarkdown("[**bold** text](https://x.jolli.ai/a*b)");
		expect(html).toBe('<div class="md-line"><a class="md-link" href="https://x.jolli.ai/a*b">**bold** text</a></div>');
	});

	it("renders ordered lists as <ol> and switches list type at a marker-style boundary", () => {
		const html = renderPrBodyMarkdown("**Steps:**\n1. Open the app\n2. Click the button");
		expect(html).toBe(
			'<div class="md-line"><strong>Steps:</strong></div><ol class="md-list"><li>Open the app</li><li>Click the button</li></ol>',
		);
	});

	it("closes an ordered list and opens an unordered one when the marker style switches", () => {
		const html = renderPrBodyMarkdown("1. first\n- second");
		expect(html).toBe('<ol class="md-list"><li>first</li></ol><ul class="md-list"><li>second</li></ul>');
	});

	it("does not double-escape text already neutralized by cli's escapeGithubWrapperTags", () => {
		// escapeGithubWrapperTags turns a literal "<details>" typed in body prose into
		// this entity form so GitHub's renderer shows it inertly (see
		// cli/src/core/SummaryPrMarkdownBuilder.ts). This renderer must undo that
		// encoding before its own escHtml pass, or the leftover "&" gets escaped a
		// second time and the browser shows the literal text "&lt;details&gt;".
		const html = renderPrBodyMarkdown("Wrap the section in a &lt;details&gt; tag");
		expect(html).toBe('<div class="md-line">Wrap the section in a &lt;details&gt; tag</div>');
		expect(html).not.toContain("&amp;lt;");
	});

	it("falls back to escaped plain text for a <summary> line that isn't the exact wrapInGithubDetails shape", () => {
		// SUMMARY_LINE is anchored to <summary><strong>…</strong></summary> specifically
		// so that untrusted content smuggled into the line (rather than pre-escaped by
		// the caller) fails the pattern and gets HTML-escaped instead of passed through.
		const html = renderPrBodyMarkdown("<summary><img src=x onerror=alert(1)></summary>");
		expect(html).not.toContain("<img");
		expect(html).toContain("&lt;summary&gt;&lt;img");
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/toolwindow/views/CreatePrBodyMarkdownTest.kt
package ai.jolli.jollimemory.toolwindow.views

import ai.jolli.jollimemory.toolwindow.views.CreatePrBodyMarkdown.renderPrBodyMarkdown
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import org.junit.jupiter.api.Test

/**
 * Kotlin mirror of `vscode/src/views/CreatePrBodyMarkdown.test.ts`. The two
 * renderers must agree line-for-line on the folding-tag whitelist: any drift is
 * a UI regression (the JCEF pane and the VS Code webview render the same
 * markdown from [SummaryPrMarkdownBuilder] side-by-side).
 */
class CreatePrBodyMarkdownTest {

    @Test
    fun `blank input returns empty string`() {
        renderPrBodyMarkdown("") shouldBe ""
        renderPrBodyMarkdown("   \n\t  ") shouldBe ""
    }

    @Test
    fun `headings offset by +1 so top-level ## renders as h3`() {
        val html = renderPrBodyMarkdown("## Summary")
        html shouldContain """<h3 class="md-heading">Summary</h3>"""
    }

    @Test
    fun `heading level clamps at h6`() {
        // `######` = 6, +1 = 7, clamp to 6. Anything deeper stops at h6.
        renderPrBodyMarkdown("###### Deep") shouldContain """<h6 class="md-heading">Deep</h6>"""
    }

    @Test
    fun `structural HTML tags pass through verbatim (Topics folding regression)`() {
        val body = """
            <details>
            <summary><strong>01 · A Topic Title</strong></summary>
            <br>
            <blockquote>

            Body text.

            </blockquote>
            </details>
        """.trimIndent()
        val html = renderPrBodyMarkdown(body)
        // Every whitelisted structural tag must appear as real HTML.
        html shouldContain "<details>"
        html shouldContain "<summary><strong>01 · A Topic Title</strong></summary>"
        html shouldContain "<br>"
        html shouldContain "<blockquote>"
        html shouldContain "</blockquote>"
        html shouldContain "</details>"
        // And none of them as escaped text (the old bug displayed &lt;details&gt;).
        html shouldNotContain "&lt;details&gt;"
        html shouldNotContain "&lt;summary&gt;"
    }

    @Test
    fun `non-whitelisted HTML is escaped defensively`() {
        // `<script>` is not in the passthrough list, so it must be escaped —
        // an LLM-generated body cannot inject arbitrary markup.
        val html = renderPrBodyMarkdown("<script>alert(1)</script>")
        html shouldContain "&lt;script&gt;alert(1)&lt;/script&gt;"
        html shouldNotContain "<script>"
    }

    @Test
    fun `summary shape must exactly match — a malformed summary line falls through to escape`() {
        // Anchored regex — an extra attribute or nested tag inside <strong> makes
        // the line fail the SUMMARY_LINE match, so it goes through the paragraph
        // branch and gets escaped. Fail-closed guarantee.
        val bad = "<summary><strong>ok<img src=x></strong></summary>"
        val html = renderPrBodyMarkdown(bad)
        html shouldContain "&lt;img src=x&gt;"
        html shouldNotContain "<img src=x>"
    }

    @Test
    fun `undoGithubWrapperEntities reverses escapeGithubWrapperTags for inline prose`() {
        // The CLI-side escapeGithubWrapperTags encodes a stray "<details>" typed in
        // body prose as "&lt;details&gt;" so GitHub renders it inertly. On the
        // display side we want the SAME "&lt;details&gt;" in the final HTML (so the
        // browser shows literal "<details>" as text) — without the undo pass we'd
        // double-escape to "&amp;lt;details&amp;gt;" and the browser would show the
        // entity text "&lt;details&gt;" instead. Verify the round-trip.
        val html = renderPrBodyMarkdown("The tag is &lt;details&gt; here.")
        html shouldContain "&lt;details&gt;"
        html shouldNotContain "&amp;lt;details&amp;gt;"
    }

    @Test
    fun `inline markdown — code, bold, italic, links`() {
        val html = renderPrBodyMarkdown("A `code` **bold** *italic* [link](https://x.y)")
        html shouldContain """<code class="md-inline-code">code</code>"""
        html shouldContain "<strong>bold</strong>"
        html shouldContain "<em>italic</em>"
        html shouldContain """<a class="md-link" href="https://x.y">link</a>"""
    }

    @Test
    fun `underscores are NOT bold — file paths and identifiers stay untouched`() {
        val html = renderPrBodyMarkdown("The var _foo_bar_ and file some_path_here.ts.")
        html shouldNotContain "<em>"
        html shouldNotContain "<strong>"
    }

    @Test
    fun `code span content is not reprocessed for bold or italic`() {
        // `**not bold**` inside a code span — the placeholder trick must keep the
        // literal asterisks out of the bold pass.
        val html = renderPrBodyMarkdown("`**not bold**`")
        html shouldContain """<code class="md-inline-code">**not bold**</code>"""
        html shouldNotContain "<strong>not bold</strong>"
    }

    @Test
    fun `unordered lists render into ul md-list with escaped item text`() {
        val html = renderPrBodyMarkdown("- item one\n- item <two>")
        html shouldContain """<ul class="md-list">"""
        html shouldContain "<li>item one</li>"
        html shouldContain "<li>item &lt;two&gt;</li>"
    }

    @Test
    fun `ordered lists render into ol md-list`() {
        val html = renderPrBodyMarkdown("1. first\n2. second")
        html shouldContain """<ol class="md-list">"""
        html shouldContain "<li>first</li>"
    }

    @Test
    fun `switching list type closes and reopens`() {
        val html = renderPrBodyMarkdown("- a\n1. b")
        // Both lists must appear; the ul closes before the ol opens.
        html shouldContain """<ul class="md-list"><li>a</li></ul><ol class="md-list"><li>b</li></ol>"""
    }

    @Test
    fun `markdown blockquote merges consecutive lines with br separators`() {
        val html = renderPrBodyMarkdown("> line 1\n> line 2")
        html shouldContain """<blockquote class="md-quote">line 1<br />line 2</blockquote>"""
    }

    @Test
    fun `horizontal rule renders md-hr`() {
        renderPrBodyMarkdown("---") shouldContain """<hr class="md-hr" />"""
        renderPrBodyMarkdown("***") shouldContain """<hr class="md-hr" />"""
    }

    @Test
    fun `fenced code block escapes body — never emits raw markup`() {
        val html = renderPrBodyMarkdown("```\n<script>bad</script>\n```")
        html shouldContain """<pre class="md-code-block"><code>&lt;script&gt;bad&lt;/script&gt;</code></pre>"""
    }

    @Test
    fun `unterminated fenced block still flushes buffered content`() {
        val html = renderPrBodyMarkdown("```\nsome text without closing fence")
        html shouldContain """<pre class="md-code-block"><code>some text without closing fence</code></pre>"""
    }

    @Test
    fun `blank line emits md-blank spacer`() {
        val html = renderPrBodyMarkdown("first line\n\nsecond line")
        html shouldContain """<div class="md-blank"></div>"""
    }
}
